Change GetParametersNoCopy to GetParametersAsSpan and use it in a few more places

stephentoub opened the pull request:
The internal GetParametersNoCopy avoids a defensive ParameterInfo[] copy, but it requires the callers to promise not to mutate the array. I'm changing the method to return a `ReadOnlySpan<ParameterInfo>`, which means a caller can't mutate it without using unsafe code, and enabling the compiler to flag any misuse.  I've then used it in a few more places. We might subsequently want to consider exposing it publicly.

jkotas commented:
Nit: `GetParametersAsSpan` may be more consistent with our naming convention for overloading on return value.

Existing art https://github.com/dotnet/runtime/blob/b049f42e5977ead2735402a79ad8247441cc3c7c/src/libraries/System.IO.Hashing/ref/System.IO.Hashing.cs#L14

jkotas reviewed (approved):
Thanks!

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.cs:
```diff
@@ -329,15 +329,10 @@ public override string ToString()
+            GetParametersAsSpan().ToArray();
 
-        internal override ParameterInfo[] GetParametersNoCopy() => LoadParameters();
+        internal override ReadOnlySpan<ParameterInfo> GetParametersAsSpan() => LoadParameters();
```
You could move LoadParameters implementation here

stephentoub replied:
I could, but it would complicate DefineParameter, which uses members defined on the derived RuntimeParameterInfos it gets back from LoadParameters.

stephentoub committed:
Change GetParametersNoCopy to GetParametersAsSpan and use it  in a few more places

The internal GetParametersNoCopy avoids a defensive ParameterInfo[] copy, but it requires the callers to promise not to mutate the array. I'm changing the method to return a `ReadOnlySpan<ParameterInfo>`, which means a caller can't mutate it without using unsafe code, and enabling the compiler to flag any misuse.  I've then used it in a few more places. We might subsequently want to consider exposing it publicly.

stephentoub committed:
Update src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs

stephentoub commented:
/azp run runtime-extra-platforms

stephentoub commented:
Both failures are known and tracked. I ran extra platforms as well, and all of those also appear to be known, with the exception of some recently added LINQ tests failing on nativeaot for which I opened an issue.
